### PR TITLE
sg-eth-airdrop

### PR DESCRIPTION
**Pull request by humanalgorithm**
Note: This PR was the parent branch for multiple admins for collection whitelist, which was needed in order for the airdrop contract to modify the collection whitelist. 

*Code Review Update*:
After the code review I made the requested changes: 
1) Changed all .unwrap( ) methods to ? operator. 
2) Validation on min and max air drop fee
3) Fairburn and instantiation fees
4) Whitelist generic was changed to whitelist-immutable, only storing addresses and counts but with no way to modify them once it has been instantiated. This module is covered with tests for the new/reduced functionality. 
5) Validation functions have all been moved to validate? methods that bubble error up to response.
6) Created a lib.rs in each folder, then did pub use::{< functions >} to export desired modules

Goal of this PR was to write the part of the airdrop in which a user gets added to a whitelist but a myriad of other changes were required: 
- copy and refactor unit tests structure from vending-minter
-  mock contracts in unit tests
- multiple admins for collection whitelist
- file structure organization

This PR will get merged into [humanalgorithm/levana-airdrop-contract-2](https://github.com/public-awesome/launchpad/tree/humanalgorithm/levana-airdrop-contract-2) which already had a working payment of airdrop funds. Adding to the collection was the complex part of this project which is why this PR got to its size.

### CODE REVIEW

File: contracts/sg-eth-airdrop/src/helpers/validation.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+pub fn validate_mints_remaining(
+    deps: &DepsMut,
+    eth_address: &str,
+) -> Result<Response, ContractError> {
+    let mint_count = ADDRS_TO_MINT_COUNT.load(deps.storage, eth_address);
+    let mint_count = mint_count.unwrap_or(0);
+    let per_address_limit = query_per_address_limit(&deps.as_ref())?;
+    if mint_count < per_address_limit {
+        Ok(Response::new())
```
Comment (shanev) [inline contracts/sg-eth-airdrop/src/helpers/validation.rs:89]: Can return just `Ok(())`

File: contracts/sg-eth-airdrop/src/helpers/validation.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+pub fn validate_eth_sig(
+    deps: &DepsMut,
+    info: MessageInfo,
+    eth_address: String,
+    eth_sig: String,
+    config: Config,
+) -> Result<Response, ContractError> {
+    let valid_eth_sig = compute_valid_eth_sig(deps, info, &config, eth_sig, eth_address.clone())?;
+    match valid_eth_sig {
+        true => Ok(Response::new()),
```
Comment (shanev) [inline contracts/sg-eth-airdrop/src/helpers/validation.rs:74]: Just `Ok(())`

File: contracts/sg-eth-airdrop/src/helpers/validation.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    Ok(Response::new())
+}
+
+pub fn validate_is_eligible(
+    deps: &DepsMut,
+    eth_address: String,
+) -> Result<Response, ContractError> {
+    let eligible = query_airdrop_is_eligible(deps.as_ref(), eth_address.clone())?;
+    match eligible {
+        true => Ok(Response::new()),
```
Comment (shanev) [inline contracts/sg-eth-airdrop/src/helpers/validation.rs:58]: Just `Ok(())` should be enough.

File: contracts/sg-eth-airdrop/src/helpers/validation.rs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    deps: &DepsMut,
+    info: MessageInfo,
+    eth_address: String,
+    eth_sig: String,
+    config: Config,
+) -> Result<Response, ContractError> {
+    validate_is_eligible(deps, eth_address.clone())?;
+    validate_eth_sig(deps, info, eth_address.clone(), eth_sig, config)?;
+    validate_mints_remaining(deps, &eth_address)?;
+    Ok(Response::new())
```
Comment (shanev) [inline contracts/sg-eth-airdrop/src/helpers/validation.rs:49]: Just `Ok(())` is fine here since this is a validation and not sending this response to the chain.

File: contracts/sg-eth-airdrop/src/helpers/constants.rs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+pub const GENERIC_WHITELIST_LABEL: &str = "Generic Whitelist for Airdrop";
+pub const NATIVE_DENOM: &str = "ustars";
+pub const CONTRACT_NAME: &str = "crates.io:sg-eth-airdrop";
```
Comment (shanev) [inline contracts/sg-eth-airdrop/src/helpers/constants.rs:3]: Weird to see something as critical as the contract name hidden inside a folder called "helpers", out of context from where it is set.

**Review by shanev — CHANGES_REQUESTED** [review]
Amazing job in getting this done! The tests are extremely thorough.

I found the code organizational structure a bit hard to follow though. Writing readable and modular code is important, but it is possible to take it to extremes where it violates the goal of [low coupling, high cohesion](https://medium.com/clarityhub/low-coupling-high-cohesion-3610e35ac4a6).

> Related code should be close to each other to make it highly cohesive

> Low cohesion would mean that the code that makes up some functionality is spread out all over your code-base. Not only is it hard to discover what code is related to your module, it is difficult to jump between different modules and keep track of all the code in your head.

I found myself having to jump around 5-6 different files to follow the logic of the contract. When organizing code, it's better to organize by feature / domain than language construct. For example all the constants for a certain feature should be in a module / file for that feature, not in a file called `constants.rs`. Same for `compute.rs`. Furthermore, the core logic is in a folder called "helpers".  In many places, functions have to be public only because they are in a different file from where they are used. In these cases, the compiler and language are giving you a clue that something is off with the organization.

File: contracts/sg-eth-airdrop/src/contract.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        eth_address.clone(),
+        eth_sig,
+        config.clone(),
+    )?;
+    let res = build_messages_for_claim_and_whitelist_add(&deps, info, config.airdrop_amount)?;
+    increment_local_mint_count_for_address(deps, eth_address)?;
+
+    Ok(res
+        .add_attribute("claimed_amount", config.airdrop_amount.to_string())
+        .add_attribute("minter_address", config.minter_address.to_string()))
```
[the remark this answers is not in the document]
Comment (humanalgorithm) [reply contracts/sg-eth-airdrop/src/contract.rs:80]: We need the claimed amount because otherwise there is no way to know how much was sent out. The minter address can otherwise be obtained from the "get_minter" function. Returning the eth address to the front end is not meaningful bc the front end is the one sending the eth address to this contract.

File: contracts/sg-eth-airdrop/src/helpers/validation.rs (hunk 1 above)
Comment (humanalgorithm) [reply contracts/sg-eth-airdrop/src/helpers/validation.rs]: Screen Shot 2022-11-28 at 12 32 35 PM
This isn't going to work, need to return the response::new( ) in each of these functions

File: contracts/sg-eth-airdrop/src/contract.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (humanalgorithm) [reply contracts/sg-eth-airdrop/src/contract.rs]: Removed minter_address from response, now getting minter address from query_minter function. Still need claimed amount so know how much was claimed.

File: contracts/vending-minter/Cargo.toml [hunk 7]
```diff
@@ -43,11 +43,11 @@ cw721           = { workspace = true }
 cw721-base      = { workspace = true, features = ["library"] }
 cw-storage-plus = { workspace = true }
 cw-utils        = { workspace = true }
-rand_core       = { version = "0.6.0", default-features = false }
+rand_core       = { workspace = true, default-features = false }
```
[the remark this answers is not in the document]
Comment (humanalgorithm) [reply contracts/vending-minter/Cargo.toml]: Removed

**Review by shanev — APPROVED** [review]
